test/unit/utils/cuda/cuda_ptr_test.cpp

```diff
@@ -0,0 +1,222 @@
+                            "Failed to cuPointerGetAttributes");
+
+    is_dev = (mem_type == CU_MEMORYTYPE_DEVICE);
+}
+
+void allocateCUDA(int dev_id, size_t len, void* &addr)
+{
+    bool is_dev;
+    CUdevice dev;
+    CUcontext ctx;
+
+    CHECK_CUDA_ERROR(cudaMalloc(&addr, len), "Failed to allocate CUDA buffer 0");
+    cudaQueryAddr(addr, is_dev, dev, ctx);
+    cout << "CUDA addr: " << std::hex << addr << " dev=" << std::dec << dev
+              << " ctx=" << std::hex << ctx << std::dec << std::endl;
+}
+
+void releaseCUDA(int dev_id, void* addr)
+{
+    CHECK_CUDA_ERROR(cudaSetDevice(dev_id), "Failed to set device");
+    CHECK_CUDA_ERROR(cudaFree(addr), "Failed to allocate CUDA buffer 0");
+}
+
+#ifdef HAVE_CUDA_VMM
+
+#define ROUND_UP(value, granularity) ((((value) + (granularity) - 1) / (granularity)) * (granularity))
+
+namespace {
+    static size_t __attribute__((unused)) padded_size = 0;
+    static CUmemGenericAllocationHandle __attribute__((unused)) handle;
+}
+
+void allocateVMM(int dev_id, size_t len, void* &_addr)
+{
+    CUdeviceptr addr = 0;
+    size_t granularity = 0;
+    CUmemAllocationProp prop = {};
+
+    CHECK_CUDA_ERROR(cudaSetDevice(dev_id), "Failed to set device");
+
+    prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+    // prop.requestedHandleTypes = CU_MEM_HANDLE_TYPE_FABRIC;
+    prop.allocFlags.gpuDirectRDMACapable = 1;
+    prop.location.id = dev_id;
+    prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    // prop.location.type = CU_MEM_LOCATION_TYPE_HOST_NUMA;
+
+    // Get the allocation granularity
+    CHECK_CUDA_DRIVER_ERROR(cuMemGetAllocationGranularity(&granularity, &prop,
+                                                          CU_MEM_ALLOC_GRANULARITY_MINIMUM),
+                            "Failed to get allocation granularity");
+    cout << "Granularity: " << granularity << std::endl;
+
+    padded_size = ROUND_UP(len, granularity);
+    CHECK_CUDA_DRIVER_ERROR(cuMemCreate(&handle, padded_size, &prop, 0),
+                            "Failed to create allocation");
+
+    // Reserve the memory address
+    CHECK_CUDA_DRIVER_ERROR(cuMemAddressReserve(&addr, padded_size,
+                                                granularity, 0, 0),
+                            "Failed to reserve address");
+
+    // Map the memory
+    CHECK_CUDA_DRIVER_ERROR(cuMemMap(addr, padded_size, 0, handle, 0),
+                            "Failed to map memory");
+
+    cout << "Address: " << std::hex << std::showbase << addr
+              << " Buffer size: " << std::dec << len
+              << " Padded size: " << std::dec << padded_size << std::endl;
+
+    // // Set the memory access rights
+    // CUmemAccessDesc access = {};
+    // access.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    // access.location.id = dev_id;
+    // access.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    // CHECK_CUDA_DRIVER_ERROR(cuMemSetAccess(addr, len, &access, 1),
```

Comment:
why commented out?

Reply:
On the system I tested this, VMM wasn't fully functional.
I.e. I could successfully run up to this step when allocating VMM, but this would fail
The steps above were enough to make VMM detectable from the cudaPtr class, but VMM can't be used.